### DISCUSSION TITLE
Middleware not working on IIS server

### DISCUSSION THREAD

**Discussion by asimplify** [vercel/next.js#49545] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64      
      Version: Windows 10 Pro
    Binaries:
      Node: 16.17.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.3.1
      eslint-config-next: 13.3.1
      react: 18.2.0
      react-dom: 18.2.0
```

### Link to the code that reproduces this issue

-SandBox

https://codesandbox.io/p/sandbox/amazing-drake-nzq90u?selection=%5B%7B%22endColumn%22%3A2%2C%22endLineNumber%22%3A9%2C%22startColumn%22%3A2%2C%22startLineNumber%22%3A9%7D%5D

-Github

https://github.com/asimplify/nextjsMiddlewareIisServer

### To Reproduce

Firstly install iisnode and UrlRewrite... this link would be helpful [ ( https://harveywilliams.net/blog/installing-iisnode  ]  then add a website in iis server. After that  simply put the app in your "C:\inetpub\wwwroot\" folder, then open cmd in the folder and run npm i followed by npm build. Then open the created website link, it should show your app running.

### Describe the Bug

The middleware doesn't to work when deployed on iis server. but it is working fine on local build and when deployed on vercel.

### Expected Behavior

Middleware should be able to redirect using my configurations.

### How are you deploying your application? (if relevant)

iis server

**Comment by JesseKoldewijn** [comment]
Hey man!
I'll have a look for ya later today when I get off work.
Do you by chance have a repo as well (besides the codesandbox) this would make setting up the reproduction a bit faster 😉

**Reply by asimplify** [reply]
https://github.com/asimplify/nextjsMiddlewareIisServer

**Reply by umashankar-pardhi-amla** [reply]
@JesseKoldewijn Any update on this ?

**Reply by JesseKoldewijn** [reply]
Sorry man, didn't get a ping from the earlier reply. I can potentially take a look at it this weekend. Currently have too busy of a schedule during the week😅

**Reply by marko-hologram** [reply]
I'm having issues with IIS and Next.js middleware as well. I'm deploying using IISNode and not using PM2 or anything else from Node ecosystem to handle the app lifecycle. IIS or specifically IISNode is responsible for starting/stopping the app.

My guess is that in my particular scenario problem is caused by iisnode using some sort of pipe instead of a port to run the app and route traffic to it.

For example this is the output I get when starting the app:

```
   ▲ Next.js 14.1.0
   - Local:        http://0.0.0.0:\\.\pipe\b65e0d78-6a8f-4375-b299-612ee9c2fc1a
   - Network:      http://\\.\pipe\b65e0d78-6a8f-4375-b299-612ee9c2fc1a:\\.\pipe\b65e0d78-6a8f-4375-b299-612ee9c2fc1a

 ✓ Ready in 271ms
```

As you can see it's not your typical `http://0.0.0.0:3000` url for example. I do have another app on another server that's being ran using PM2 and IIS is just using Application Request Routing module to proxy to that app and middleware works fine there. The only difference seems to be that the app there is started on a normal port.

Everything else works fine, but middleware just doesn't run at all.

**Reply by vaishnavi-jadhav-amla** [reply]
any update on this issue. I am facing similar issue

**Reply by marko-hologram** [reply]
@vaishnavi-jadhav-amla you might need to provide more details. How are you running your app on IIS, how are you starting it? Do you have custom server file for example etc.

---

I personally stopped using IIS (IISNode) directly to start the app and just run the app using PM2 on the server and use IIS module called Application Request Routing to reverse proxy to the application running on a certain port like 3000 for example.

**Comment by Smirra** [comment]
It is actually possible to freely be able to utilise all the functionality of the middleware. Atleast as far as I have managed to test it.

When using the middleware without any matcher, it will be invoked for **every route** in your project. For some reason the named pipe makes it all the way into NextRequest and is thus prepended to all the paths on the form:
> 0.0.0.0/**/pipe/\<uuid>**/\<optional-baseUrl>/\<the-rest-of-the-paths>

and what Next.js wants to receive from the NextResponse is:
> 0.0.0.0/\<optional-baseUrl>/\<the-rest-of-the-paths>

So the confusing parts were first realizing that it prepends the named pipe on all the NextRequests, which means you simply cannot filter out all the calls beginning with "//pipe", and second that the pipe segment shouldn't be forwarded further through the NextResponse.

Since you have to do kind of 'a lot of magic' to each request that you actually want to rewrite or redirect, I found it easier to handle all the logic conditionally and completely opt out of the matcher and any redirect/rewrites in the next.config.js. `nextUrl.basePath` returns an empty string probably because of the named pipe prepending, so you have to save it globally somewhere. Not sure if you even need to set the basePath in next.config.js when taking over this much of the logic. But better be safe than sorry. Note that just invoking NextResponse.next() still works, so doing something other than redirecting and rewriting is already working.

Basic example on how to make redirect/rewrite work:
```
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  const url = request.nextUrl
  const pathname = url.pathname

  // Match on all the static stuff directly and move on
  if (pathname.match('/(_next/static|_next/image|favicon.ico|sitemap.xml|robots.txt).*')) {
    return NextResponse.next()
  }

  const parts = pathname.split('/').filter(Boolean)
  let basePathSegment = '', rest = ''

  // Splice away the pipe junk
  if (parts[0] === 'pipe' && parts.length > 1) {
    parts.splice(0, 2)
  }

  // Splice away the base path and save it
  if (parts.length > 0 && `/${parts[0]}` === process.env.BASEPATH) {
    basePathSegment = `/${parts[0]}`
    parts.splice(0, 1)
  }

  rest = `/${parts.join('/')}`

  // Insert some redirect or rewrite conditional logic to return here
  // and add together the basePathSegment + someMagicBefore + rest + someMagicAfter,
  // or whatever you want to do with the rest instead of joining them as they came in,
  // and return it

  // Do nothing and move on if no redirect / rewrite condition was met
  return NextResponse.next()
}
```

```
Operating System:
      Platform: win32
      Arch: x64   
      Version: Windows 10 Pro
    Binaries:
      Node: 16.17.1
      npm: 10.8.3
      Yarn: N/A
      pnpm: N/A
      IIS: 10
    Relevant packages:
      next: 14.1.4
      eslint-config-next: 14.1.4
      react: 18.2.0
      react-dom: 18.2.0
```
